### PR TITLE
Do not create a cluster if it already exists.

### PR DESCRIPTION
**Pull request by ehiggs**

### CODE REVIEW

**Comment by ehiggs** [comment]
Jenkins: test this please.

**Comment by ehiggs** [comment]
Jenkins: test this please.

File: hod/subcommands/create.py [hunk 1]
```diff
@@ -83,10 +85,24 @@ def run(self, args):
             sys.stderr.write('Missing config options. Exiting.\n')
             return 1
 
+        label = optparser.options.label
+
+        if label is not None and not hc.is_valid_label(label):
+            sys.stderr.write("Tried to submit HOD cluster with label '%s' but it is not a valid label\n" % label)
+            sys.stderr.write("Labels are used as filenames so they cannot have %s characters\n" % os.sep)
+            sys.exit(1)
```
Comment (boegel) [inline hod/subcommands/create.py]: don't copy-paste this, create a `check_valid_label` function and call it

File: hod/subcommands/create.py [hunk 2]
```diff
@@ -83,10 +85,24 @@ def run(self, args):
...
+
+        if label in hc.known_cluster_labels():
+            sys.stderr.write("Tried to submit HOD cluster with label '%s' but it already exists\n" % label)
+            sys.stderr.write("If it is an old HOD cluster, you can remove it with `hod clean`\n")
+            sys.exit(1)
```
Comment (boegel) [inline hod/subcommands/create.py]: likely, no copy-pasta puh-leez

File: hod/subcommands/create.py [hunk 3]
```diff
@@ -83,10 +85,24 @@ def run(self, args):
...
+
         try:
-            # FIXME: check whether label is already in use
             j = PbsHodJob(optparser)
-            print "Submitting HOD cluster with label '%s'..." % optparser.options.label
+            if label is None:
+                print "Submitting HOD cluster with no label ..."
```
Comment (boegel) [inline hod/subcommands/create.py]: mention that job ID will be used as default label?

File: hod/subcommands/batch.py [hunk 4]
```diff
@@ -93,6 +95,17 @@ def run(self, args):
             sys.stderr.write('Missing script. Exiting.\n')
             return 1
 
+        label = optparser.options.label
+
+        if label is not None and not hc.is_valid_label(label):
+            sys.stderr.write("Tried to submit HOD cluster with label '%s' but it is not a valid label\n" % label)
+            sys.stderr.write("Labels are used as filenames so they cannot have %s characters\n" % os.sep)
+
+        if optparser.options.label in hc.known_cluster_labels():
```
Comment (boegel) [inline hod/subcommands/batch.py]: `known_cluster_labels()` doesn't include labels for clusters that are submitted but not running yet, so it's still possible to reuse a label (by accident)

can we avoid this, by making `known_cluster_labels()` also check submitted-but-not-yet-running jobs, or making `hod create` create the cluster info dir rather than having `local` do it (we'll need to be careful with `hod clean`, e.g. only clean up a cluster if a job ID is known for it?)

File: hod/cluster.py [hunk 5]
```diff
@@ -63,6 +64,30 @@
...
+    directory splitting characters in it.
+
+    None is also a value label; it just means we will use the job id for the
+    label.
+    """
+    return path is None or os.path.basename(path) == path
+
+
+def validate_label(label, known_labels):
+    if not is_valid_label(label):
```
Comment (boegel) [inline hod/cluster.py]: docstring pls

File: hod/subcommands/create.py [hunk 6]
```diff
@@ -83,10 +85,17 @@ def run(self, args):
...
+            sys.exit(1)
+
         try:
-            # FIXME: check whether label is already in use
             j = PbsHodJob(optparser)
-            print "Submitting HOD cluster with label '%s'..." % optparser.options.label
+            if label is None:
+                print "Submitting HOD cluster with no label (job id will be used as a default label) ..."
+            else:
+                print "Submitting HOD cluster with label '%s'..." % label
```
Comment (boegel) [inline hod/subcommands/create.py]: more copy-paste (see batch.py)

looks like a big part of both create.run and batch.run should be hoisted in a separate function?

File: hod/subcommands/batch.py [hunk 7]
```diff
@@ -93,8 +95,18 @@ def run(self, args):
             sys.stderr.write('Missing script. Exiting.\n')
             return 1
 
+        label = optparser.options.label
+
+        if not hc.validate_label(label, hc.known_cluster_labels()):
```
Comment (boegel) [inline hod/subcommands/batch.py]: my remark still stands: `known_cluster_labels()` does not include labels for HOD clusters that are submitted but not running yet?

Comment (ehiggs) [inline hod/subcommands/batch.py]: Fixed. `hod list` now shows the job id (which has the cluster) and the pbs job state. e.g.

```
$ hod list
Cluster label            	Job ID                            	PBS Job State
this-is-my-delcatty-label	1736910.master15.delcatty.gent.vsc	Jobid 1736910.master15.delcatty.gent.vsc state R ehosts node2127.delcatty.gent.vsc
this-is-my-label         	100814.master20.golett.gent.vsc   	<job-not-found>
my-batch-job             	100810.master20.golett.gent.vsc   	<job-not-found>
```

File: hod/cluster.py [hunk 8]
```diff
@@ -63,6 +64,45 @@
...
 
+def is_valid_label(path):
+    """
+    Checks if a labelname is a valid filename by making sure it doesn't have
+    directory splitting characters in it.
+
+    None is also a value label; it just means we will use the job id for the
+    label.
+    """
+    return path is None or os.path.basename(path) == path
```
Comment (boegel) [inline hod/cluster.py]: `None` is a valid label?

File: hod/cluster.py [hunk 9]
```diff
@@ -63,6 +64,45 @@
 
 ClusterInfo = namedtuple('ClusterInfo', 'label, jobid, pbsjob')
 
+def is_valid_label(path):
```
Comment (boegel) [inline hod/cluster.py]: s/path/label/?

File: hod/cluster.py [hunk 10]
```diff
@@ -177,9 +223,22 @@ def save_cluster_info(cluster_info):
...
+        with open(os.path.join(info_dir, 'jobid'), 'w') as jobid_file:
+            jobid_file.write(jobid)
+    except IOError as err:
+        _log.error("Failed to write cluster info files: %s", err)
+
+
+def save_cluster_info(cluster_info):
+    """Save info (job ID, env script, ...) for this cluster in the cluster info dir."""
+    info_dir = os.path.join(cluster_info_dir(), cluster_info['label'])
+    jobid = os.getenv('PBS_DEFAULT', 'PBS_JOBID_NOT_DEFINED')
```
Comment (boegel) [inline hod/cluster.py]: `s/PBS_DEFAULT/PBS_JOBID/g`?

This looks very wrong/untested?

File: hod/subcommands/batch.py [hunk 11]
```diff
@@ -93,11 +95,18 @@ def run(self, args):
...
+            sys.exit(1)
+
         try:
             j = PbsHodJob(optparser)
+            hc.report_cluster_submission(label)
             j.run()
             jobs = j.state()
             print "Jobs submitted: %s" % [str(j) for j in jobs]
+            hc.mk_cluster_info(label, jobs[0].jobid)
```
Comment (boegel) [inline hod/subcommands/batch.py]: since you're using `[0]` here, make very sure that `jobs` is a single-element list, and go down hard if it's not (and try to remove any jobs that may have been submitted)?

File: hod/subcommands/create.py [hunk 12]
```diff
@@ -83,13 +85,18 @@ def run(self, args):
...
+
         try:
-            # FIXME: check whether label is already in use
             j = PbsHodJob(optparser)
-            print "Submitting HOD cluster with label '%s'..." % optparser.options.label
+            hc.report_cluster_submission(label)
             j.run()
             jobs = j.state()
             print "Jobs submitted: %s" % [str(j) for j in jobs]
+            hc.mk_cluster_info(label, jobs[0].jobid)
```
Comment (boegel) [inline hod/subcommands/create.py]: same here, and again, duplicate code?

File: hod/subcommands/listcmd.py [hunk 13]
```diff
@@ -48,10 +48,11 @@ class ListOptions(GeneralOption):
     """Option parser for 'list' subcommand."""
     VERSION = HOD_VERSION
 
-def mk_fmt_str(labels, header):
+def mk_fmt_str(labels, jobid, header):
     """Calculate the format string for printing the template parameters."""
     max_label_len = max(map(len, chain(labels, [header])))
-    return '%%-%ds\t%%s' % max_label_len
+    max_jobid_len = max(map(len, chain(jobid, [header])))
+    return '%%-%ds\t%%-%ds\t%%s' % (max_label_len, max_jobid_len)
```
Comment (boegel) [inline hod/subcommands/listcmd.py]: this is becoming pretty much impossible to comprehend, can we make this more readable, e.g. by using `'foo' + '%s' + 'bar%s'` style?

File: hod/subcommands/listcmd.py [hunk 14]
```diff
@@ -59,8 +60,9 @@ def format_cluster_info(cluster_info):
     ret = []
     for info in cluster_info:
         label = info.label if info.label is not None else '<no-label>'
-        job = str(info.pbsjob) if info.pbsjob is not None else '<no-job>'
-        ret.append((label, job))
+        jobid = info.jobid
+        job = str(info.pbsjob) if info.pbsjob is not None else '<job-not-found>'
```
Comment (boegel) [inline hod/subcommands/listcmd.py]: why the change from `no-job`?

Comment (ehiggs) [inline hod/subcommands/listcmd.py]: Because there is ostensibly a job as concept; but it isn't found on the server.

File: hod/cluster.py (hunk 8 above)
Comment (ehiggs) [inline hod/cluster.py]: Yes, if someone passes no label, it is valid..

Comment (boegel) [inline hod/cluster.py]: No, then we use `$PBS_JOBID`, *that's* a valid label. `None` is not.

Comment (ehiggs) [inline hod/cluster.py]: Yes but this is used in parsing the command line input to make sure a valid label was given. As opposed to "my/super/label" which wouldn't work. We don't know the job name. Obviously the docstring is lacking as it wasn't clear to you what this is for.
